apps/emqx_durable_storage/src/emqx_ds_cache.erl

```diff
@@ -0,0 +1,513 @@
+        Iterators0
+    ).
+
+-spec handle_pull(state()) -> state().
+handle_pull(State0) ->
+    #{
+        db := DB,
+        iterators := Iterators0,
+        batch_size := BatchSize
+    } = State0,
+    Iterators =
+        maps:fold(
+            fun(TopicFilter, StreamsToIt0, Acc) ->
+                StreamsToIt = do_pull(DB, BatchSize, StreamsToIt0),
+                Acc#{TopicFilter := StreamsToIt}
+            end,
+            Iterators0,
+            Iterators0
+        ),
+    State0#{iterators := Iterators}.
+
+-spec do_pull(emqx_ds:db(), pos_integer(), streams()) -> streams().
+do_pull(DB, BatchSize, StreamsToIt) ->
+    maps:fold(
+        fun
+            (_Stream, #{it := ?EOS}, Acc) ->
+                Acc;
+            (Stream, #{it := It0, seqno := Seqno}, Acc) ->
+                {It, NewSeqno} = do_pull1(DB, BatchSize, Stream, It0, Seqno),
+                Acc#{Stream => #{it => It, seqno => NewSeqno}}
+        end,
+        StreamsToIt,
+        StreamsToIt
+    ).
+
+-spec do_pull1(emqx_ds:db(), Seqno, emqx_ds:stream(), emqx_ds:iterator(), pos_integer()) ->
+    {emqx_ds:iterator(), Seqno} | {?EOS, undefined}
+when
+    Seqno :: pos_integer().
+do_pull1(DB, BatchSize, Stream, It0, Seqno) ->
+    case emqx_ds:next(DB, It0, BatchSize, #{use_cache => false}) of
+        {ok, It, Batch} ->
+            Now = now_ms(),
+            NewSeqno = store_batch(DB, Now, Seqno, Stream, Batch),
+            {It, NewSeqno};
+        {ok, end_of_stream} ->
+            Now = now_ms(),
+            insert_end_of_stream(DB, Now, Seqno, Stream),
+            {?EOS, undefined};
+        {error, _Error} ->
+            %% log?
+            {It0, Seqno}
+    end.
+
+-spec handle_add_cached_topic_filters([emqx_ds:topic_filter()], state()) -> state().
+handle_add_cached_topic_filters(TopicFilters, State0) ->
+    #{db := DB, iterators := Iterators0} = State0,
+    NewTopicFilters = TopicFilters -- maps:keys(Iterators0),
+    NewIterators0 = maps:from_keys(NewTopicFilters, #{}),
+    NewIterators = renew_streams(DB, NewIterators0),
+    Iterators = maps:merge(Iterators0, NewIterators),
+    State0#{iterators := Iterators}.
+
+-spec handle_remove_cached_topic_filters([emqx_ds:topic_filter()], state()) -> state().
+handle_remove_cached_topic_filters(TopicFilters, State0) ->
+    #{iterators := Iterators0} = State0,
+    Iterators = maps:without(TopicFilters, Iterators0),
+    State0#{iterators := Iterators}.
+
+-spec handle_gc(state()) -> ok.
+handle_gc(State) ->
+    #{db := DB, gc_interval := GCInterval} = State,
+    Now = now_ms(),
+    MS = gc_ms(Now, GCInterval),
+    NumDeleted = ets:select_delete(?cache(DB), MS),
+    ?tp(ds_cache_gc_ran, #{num_deleted => NumDeleted}),
+    ok.
+
+gc_ms(Now, Threshold) ->
+    ets:fun2ms(fun(#entry{inserted_at = InsertedAt}) when
+        InsertedAt + Threshold =< Now
+    ->
+        true
+    end).
+
+now_ms() ->
+    erlang:system_time(millisecond).
+
+start_timer(Event, Timeout) ->
```

Comment:
nit: `emqx_utils:start_timer`

Reply:
that one wraps the message with `{timeout, TRef, _}`